<center>
<img src="https://raw.githubusercontent.com/FUlyankin/r_probability/master/end_seminars_2020/sem08/real_expect.png" width="800">

# Yet another matstat course: домашнее задание 3
</center>


## Задача 1: метрика плохих показов (13 баллов)

Мало того, что в интернете постоянно кто-то не прав, так ещё и куча спама, фейков, хейтспича, кликбейта и другого «плохого» контента. Каждая уважающая себя платформа борется с ним. Более того, гиганты вроде Facebook публикуют [transparency-отчёты](https://transparency.fb.com/reports/community-standards-enforcement/) о том, сколько показов «плохого» контента пропустила их система модерации. 

Давайте представим себе, что мы Youtube. Мы хотим, чтобы пользователям как можно реже показывался «плохой» контент (например, спам или порно). 

К сожалению, мы не можем сделать ручную разметку всех видео, которые загружают люди. Поток нового видео на платформе постоянно растёт. Нам надо будет поддерживать огромный штат модераторов. Поэтому для каждого нарушения, с помощью машинного обучения, обычно делают классификаторы. Нейросети пытаются предсказать, есть ли в видео неприемлимый контент. Если его вероятность высокая, видео автоматически банится. Если нейросеть не уверена, видео отправляют на разметку модераторам. Если мы обучим хороший классификатор, подавляющая часть потока будет оставаться без модерации. 

Параллельно с классификатором мы можем начать размечать модераторами жалобы и самые вирусные видосы. Машинное обучение даёт осечки. Какое-то одно «плохое» видео, пропущенное нашей системой, может набрать много показов. Постмодерация самых вирусных видео и жалоб могут нас от этого спасти. 

Нам хотелось бы понимать, насколько хорошо работает система модерации. Для этого мы будем оценивать долю «плохих» показов. Если мы показываем видео со спамом, такие показы мы считаем плохими.

**Наша задача**  —  получить несмещённую оценку доли плохих показов на youtube, а также построить для неё доверительный интервал. Тогда мы будем видеть, с каким нарушением у нас больше всего проблем, а также сможем придумывать для системы модерации разные улучшения и понимать, насколько они эффективны.

#### 1. Данные

У нас есть данные о просмотре видео на youtube за сутки. Таблица весьма громадная, так как она содержит несколько миллиардов уникальных видео. Мы хотим выбрать из неё тысячу случайных видео, чтобы модераторы разметили их на спам. При этом, нам нужно учесть частоту просмотров. Популярные видео должны иметь больше шансов попасть в подвыборку. Тогда по этой разметке мы сможем оценить долю плохих показов. 

|      video                         |            shows                  |
|:----------------------------------:|:---------------------------------:|
| Baby Shark                         |  13 840 000 000                   |
| Despacito                          |   8 340 000 000                   |
| Johny Johny Yes Papa               |   6 850 000 000                   |
| ........                           |   ........                        |
| 1. Андан-2025: симуляции           |   2                               |

Давайте посчитаем для каждого видео вероятность, что оно будет показно и сделаем `np.random.choice`, как в коде ниже.

In [108]:
import numpy as np 
import pandas as pd 
import scipy.stats as sts

In [109]:
# сгенерируем показы из распределения Парето, чтобы было побольше выбросов как в жизни.
n_obs = 10**6

b = 0.7
rv = sts.pareto(b)
shows = np.round(rv.rvs(n_obs))
shows # это показы

array([ 11., 776.,   2., ...,   8.,   1.,   3.])

In [110]:
videos = np.arange(1, n_obs + 1) # это id видео
videos

array([      1,       2,       3, ...,  999998,  999999, 1000000])

In [111]:
p = shows / shows.sum() # это вероятность того что видео будет показано
p

array([1.68220806e-09, 1.18672132e-07, 3.05856011e-10, ...,
       1.22342405e-09, 1.52928006e-10, 4.58784017e-10])

In [112]:
# генерируем выборку без повторений из 100 видео с учётом показов (более частые видео попадут в выборку вероятнее)
k = 100
np.random.choice(videos, size=k, replace=False, p = p)

array([134147,  29943, 502975, 839504, 815399, 283236,  63180,  32692,
       816733, 137968, 495367, 508847, 131787, 801474, 490250, 673973,
       981411, 326145, 594192, 961401, 375641, 491990, 144744,  33640,
       699164, 891474,  61462, 256898, 968547, 889200, 203417, 193528,
       310344, 281175, 563920, 432290, 215612, 542018, 869664, 525149,
       886746, 166649, 480965, 961076, 480579, 727940, 662688, 465053,
       481699,  33074, 829207, 759898, 764184,  17586, 647342, 907572,
       583561,  92122, 914275, 406202, 893923, 600901, 575639, 189960,
       232370, 344525, 464087, 894465, 840662, 535412, 875492, 833772,
       716203, 282050, 347426,  26051, 423216, 390863,  61559, 419578,
       659900, 821461, 683876, 623005, 574581,  55005, 176082,   6110,
        76896, 870158, 610775, 612110, 993521, 309299,  73296, 430138,
       609113, 254664, 313001, 589589])

У такого подхода есть проблема. Если в таблице миллиарды строк, мы не сможем сохранить таблицу в оперативную память. Нам для генерации выборки понадобится супер-компьютер. Хотелось бы этого избежать. К счастью, для решения этой проблемы есть [много алгоритмов,](https://en.wikipedia.org/wiki/Reservoir_sampling) и мы с вами реализуем один из них.

#### 2. Сэмплирование с повторениями

Для начала поработаем с сэмплированием с повторениями. Сгенерируем таблицу с данными. 



In [113]:
np.random.seed(42)
n_obs = 10**5

b = 0.7
rv = sts.pareto(b)

df = pd.DataFrame.from_dict({
    'video_id': np.arange(1, n_obs + 1),
    'shows': np.round(rv.rvs(n_obs)),
    'is_spam': np.random.binomial(1, 0.10, n_obs), # пусть в 10% видео встречается спам
})

print(df.shape)
df.head()

(100000, 3)


video_id  shows  is_spam
0         1    2.0        0
1         2   74.0        0
2         3    7.0        0
3         4    4.0        0
4         5    1.0        0

__а) [1 балл]__  Посчитайте по табличке `df` истиную долю плохих показов: 

$$
p_{\text{bad}} = \frac{\sum_{v \in V} \text{show}(v) \cdot \text{isSpam}(v) }{\sum_{v \in V} \text{show}(v)}.
$$

In [114]:


### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz
# сумма произведений количества показов и метки спама
numerator = (df['shows'] * df['is_spam']).sum()

# общая сумму показов
denominator = df['shows'].sum()

# доля плохих показов
p_bad = numerator / denominator

print("Доля плохих показов:", p_bad)


Доля плохих показов: 0.030451003751416496


Мы не знаем всех меток $\text{isSpam}(v)$. Мы можем позволить себе разметить маленький сэмпл, $S ⊂ V$. Каждое видео попадает к нам в сэмпл пропорционально числу его показов. Поэтому логично оценить долю плохих показов как 

$$
\hat{p}_{\text{bad}} = \frac{1}{|\text{S}|}\sum_{v \in S} \text{isSpam}(v).
$$

__б) [1 балл]__ С помощью функции `np.random.choice` cделайте $10^4$ выборок __с повторениями__ размера $1000$. 

Постройте для каждой оценку доли плохих показов. Считайте, что модератор во время разметки безошибочно определяет значение из колонки `is_spam`. 

Нарисуйте гистограмму для получившегося распределения. Отметьте на ней настоящую долю плохих показов и получившееся у вас среднее. Правда ли, что мы получили несмещённую оценку?

**Hint:** Перейдите от `pandas` к `numpy` и обратите внимание на команду `argsort()`. Такой код будет работать быстрее.

In [115]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz

import plotly.graph_objects as go

# параметры сэмплирования
n_samples = 10**4
sample_size = 1000

spam_labels = df['is_spam'].values
shows = df['shows'].values

# Сортируем индексы по убыванию просмотров
sorted_indices = np.argsort(-shows)
sorted_shows = shows[sorted_indices]
cum_shows = np.cumsum(sorted_shows)
total_shows = cum_shows[-1]

p_hat_arr = np.empty(n_samples)

for i in range(n_samples):
    rand_vals = np.random.rand(sample_size) * total_shows
    pos = np.searchsorted(cum_shows, rand_vals, side='right')
    sample_idx = sorted_indices[pos]
    p_hat_arr[i] = spam_labels[sample_idx].mean()

# Истинная доля
true_p = (shows * spam_labels).sum() / shows.sum()

# построим гистограму 
fig = go.Figure()
fig.add_trace(go.Histogram(x=p_hat_arr, nbinsx=30, marker_color='skyblue'))
fig.add_vline(x=true_p, line=dict(color='red', dash='dash'),
              annotation_text=f'Истинная доля = {true_p:.3f}', annotation_position="top left")
fig.add_vline(x=np.mean(p_hat_arr), line=dict(color='green', dash='dash'),
              annotation_text=f'Средняя оценка = {np.mean(p_hat_arr):.3f}', annotation_position="top right")
fig.update_layout(
    title=f'Средняя оценок: {np.mean(p_hat_arr):.4f}, Истинное значение: {true_p:.4f}',
    xaxis_title='Оценка доли спама',
    yaxis_title='Частота'
)
fig.show()




#### 3. Сэмплирование без повторений

Выше мы сказали, что делать `np.random.choice` для огромных таблиц невозможно, так как они не поместятся в оперативную память. 

Более того, нам надо дать модераторам разметить видео на спам. Многие видео будут повторяться. Уникальных видео каждый раз будет разное количество. Нагрузка на модераторов будет неравномерной. Они будут жаловаться на это. Хочется, чтобы нагрузка всегда была одинаковой.

Давайте сменим подход и будем делать **случайную взвешенную выборку БЕЗ ПОВТОРЕНИЙ.** Это означает, что наблюдения зависят друг от друга. Причём корреляция между ними довольно высокая. Это будет приводить к проблемам. 

Пусть у нас есть $n$ объектов. Мы хотим отобрать $k$ видео для разметки модераторами. Использование `np.random.choice` можно проинтерпретировать следующим образом: 

1. Напишем название $i-$го видео на разных табличках $\text{shows}_i$ раз.
2. Случайно перемешаем все таблички и положим их в стопку в случайном порядке.
3. Будем отбирать самые верхние таблички до тех пор, пока не встретим $k$ разных названий видео.

Если бы нам были бы не важны веса в виде показов, каждое видео попадало бы к нам в выборку равновероятно. Мы могли бы отобрать выборку размера $k$ следущим образом: 

1. Для каждого видео генерируем $X_i \sim U[0; 1]$ (по одной табличке на видео).
2. Сортируем все видео по сгенерированной случайной величине.
3. Срезаем топ-k видео в итоговую выборку.

Веса в виде показов для нас важны. В теринах равномерных случайных величин алгоритм с весами можно записать так: 

1. Для $i$-го видео генерируем $\text{shows}_i$ независимых равномерных случайных величин (для каждого видео свое число табличек). 
2. Складываем все таблички в одну стопку и сортируем их по сгенерированным величинам.
3. Идём по массиву от больших элементов к меньшим и отбираем видео, пока не накопим $k$ элементов.

Очень не хочется для какого-нибудь популярного видео генерировать несколько миллиардов случайных чисел. Конечно же, сразу нужно генерировать случайную величину, которая будет максимумом нескольких независимых равномерных случайных величин. На итоговую сортировку это никак не повлияет. Мы отберём те же самые видео.  



Сгенерировать для каждого видео $X_{i,max}$ можно с помощью квантильного преобразования. 

Пусть случайные величины $X_1, \ldots, X_m \sim \text{iid} \, U[0;1].$ Пусть $Y = \max(X_1, \ldots, X_m).$ 

Мы знаем, что $F_{Y}(x) = x^m,$ если $x \in [0; 1].$ Выборку из распределения $F_{Y}(x)$ можно сгенерировать в два шага: 

- $x_1, \ldots, x_m \sim \text{iid} \, U[0;1]$
- $y_i = x_i^{\frac{1}{m}}$

__в) [1 балл]__ Фактически нам надо отсортировать все видео по величине $X_i^{\frac{1}{\text{shows}_i}}$, где $X_i \sim U[0;1]$, и отобрать топ-$k$ видео в выборку для разметки модераторами. Сделайте, используя эту процедуру, для таблицы `df` сэмпл размера $100$.

In [116]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz
# Генерируем ключ для каждого видео по формуле score = U^(1/shows)

X = np.random.uniform(size=df.shape[0])
df['score'] = X ** (1 / df['shows'])

sample_df = df.sort_values('score', ascending=False).head(100)

print("Выбранные видео для разметки (первые 5 строк):")
sample_df.head()


Выбранные видео для разметки (первые 5 строк):


video_id       shows  is_spam  score
15618     15619    778644.0        0    1.0
89529     89530  19257035.0        0    1.0
43676     43677   3216133.0        0    1.0
531         532    117586.0        0    1.0
58232     58233   1066786.0        0    1.0

__г) [1 балл]__ Вычислять на компьютере корни из числа, лежащего между 0 и 1 не очень удобно с точки зрения округления. Могут возникать большие ошибки.

Гораздо эффективнее отобрать топ по величине $\frac{1}{\text{shows}_i} \cdot \ln X_i$, где $X_i \sim U[0;1]$. Проделайте это.

In [117]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz
# Генерируем новый ключ с использованием логарифма

X = np.random.uniform(size=df.shape[0])
df['score_log'] = np.log(X) / df['shows']

sample_df_log = df.sort_values('score_log', ascending=False).head(100)

print("Выбранные видео для разметки (топ-100 по методу ln):")
sample_df_log.head()


Выбранные видео для разметки (топ-100 по методу ln):


video_id       shows  is_spam     score     score_log
76479     76480        17.0        0  0.997099 -3.685948e-08
89529     89530  19257035.0        0  1.000000 -3.755542e-08
53043     53044     86342.0        1  0.999997 -9.058626e-08
96707     96708   3972215.0        0  1.000000 -1.977015e-07
4787       4788        29.0        0  0.991083 -2.510191e-07

Такую процедуру можно найти в продакшн-процессах у многих компаний. В Яндексе похожая процедура используется для сэмплирования поисковых запросов для дальнейшей разметки. [В статье](https://yadi.sk/i/IxKLPFEj3TSpPe) можно найти доказательство того, что алгоритм даст корректную вероятность для каждого видео.

В примере выше, мы держим табличку в оперативной памяти компьютера. Это игрушечный пример, и это позволительно. В реальной жизни мы можем считывать строки с жесткого диска, для каждой из них по очереди генерировать случайную величину и поддерживать топ-к уникальных видео в памяти. Например, в этом может помочь такая структура данных [как куча.](https://ru.wikipedia.org/wiki/Куча_(структура_данных)) 

#### 4. Оценка доли плохих показов

Дальше нам остаётся разметить видео и аккуратно посчитать итоговую метрику. Беда будет в том, что она окажется смещённой. 

__д) [1 балл]__ Сделайте $10^4$ выборок размера $1000$. Постройте для каждой оценку доли плохих показов. Считайте, что модератор во время разметки безошибочно определяет значение из колонки `is_spam`. 

Нарисуйте гистограмму для получившегося распределения. Отметьте на ней настоящую долю плохих показов и получившееся у вас среднее. Правда ли, что они сильно отличаются друг от друга? Найдите среднее смещение оценки.

In [118]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz
# Параметры для симуляции
n_simulations = 10**4  
sample_size = 1000     
n_obs = df.shape[0]    

# Предварительно извлечём нужные массивы
spam_labels = df['is_spam'].values
shows = df['shows'].values

# Массив для хранения оценок доли спама для каждой симуляции
p_hat = np.empty(n_simulations)


# и считаем среднее по is_spam
for i in range(n_simulations):
    u = np.random.uniform(size=n_obs)
    score_log = np.log(u) / shows
    top_indices = np.argpartition(score_log, -sample_size)[-sample_size:]
    p_hat[i] = spam_labels[top_indices].mean()

true_p = (df['shows'] * df['is_spam']).sum() / df['shows'].sum()

mean_estimate = p_hat.mean()
bias = mean_estimate - true_p

print("Истинная доля плохих показов:", true_p)
print("Средняя оценка по симуляциям:", mean_estimate)
print("Среднее смещение оценки:", bias)

# Гистограмма распределения оценок
fig = go.Figure()
fig.add_trace(go.Histogram(x=p_hat, nbinsx=30, marker_color='skyblue'))
fig.add_vline(x=true_p, line=dict(color='red', dash='dash'),
              annotation_text=f'Истинная доля = {true_p:.3f}', annotation_position="top left")
fig.add_vline(x=mean_estimate, line=dict(color='green', dash='dash'),
              annotation_text=f'Средняя оценка = {mean_estimate:.3f}', annotation_position="top right")
fig.update_layout(
    title=f'Средняя оценок: {mean_estimate:.4f}, Истинное значение: {true_p:.4f}',
    xaxis_title='Оценка доли спама',
    yaxis_title='Частота'
)
fig.show()




Истинная доля плохих показов: 0.030451003751416496
Средняя оценка по симуляциям: 0.09994230000000001
Среднее смещение оценки: 0.06949129624858352


**Откуда появляется это смещение?** Представим себе, что у нас есть случайная величина $X$, которая принмает пять значений с вероятностями 

<center>

|  $X$         | $x_1$   | $x_2$ | $x_3$ | $x_4$    | $x_5$    |
|:------------:|:-------:|:-----:|:-----:|:--------:|:--------:|
| $P(X = x)$   | $^1/_2$ |$^1/_4$|$^1/_8$|$^1/_{16}$|$^1/_{16}$|

</center>

Если мы делаем выборку с повторениями, как часто туда будет попадать элемент $x_1$? Элемент попадает в выборку с вероятностью $^1/_2.$ Будем вытаскивать из выборки элементы до тех пор, пока $x_1$ не окажется в наших руках. Номер попытки, начиная с которой $x_1$ окажется у нас, имеет геометрическое распределение. Если $Y \sim \text{Geom}(p),$ тогда $\mathbb{E}(Y) = \frac{1}{p}.$ 

Получается, элемент $x_1$ окажется в нашей выборке в среднем на второй попытке. Если бы мы делали выборку с повторениями, каждый второй элемент в ней был бы равен $x_1,$  каждый четвёртый был бы равен $x_2$, каждый восьмой был бы равен $x_3$ и так далее. Обратим внимание, что если мы делаем выборку из трёх элементов без повторений, то чаще всего мы будем работать с выборкой $x_1, x_2, x_3$. 

В ситуации с видео, мы сэмплируем их пропорционально числу показов. В числе показов может быть очень сильный перекос. Какие-то видео показываются в рекомендательной системе десятки раз, а какие-то вирусятся и прорываются в тренды. Каждый раз, когда мы __без возвращения__ берём новое видео в выборку, мы как бы занижаем его вес в несмещённой выборке (выборке с повторениями).

#### 5. Исправляем смещение

На выборку без повторений размера $n$ можно смотреть следующим образом: мы генерируем выборку с возвращением до тех пор, пока количество уникальных элементов не достигнет числа $n$. При этом, если мы берём на каком-то шаге элемент уже выбранный ранее, мы не включаем его в выборку, а только отдельно запоминаем где-нибудь счётчик числа вхождений этого элемента $c_i$. При таком подходе несмещённую оценку доли плохого можно записать как 

$$
\hat{p}_{\text{bad}} = \frac{\sum_{i=1}^{n} c_i \cdot \text{isSpam}(v_i)}{\sum_{i=1}^{n} c_i.}
$$

Если выборка имеет сильно неравномерные веса, то нам придётся довольно долго генерировать элементы с возвращением, пока мы наберём необходимое количество уникальных. 

Поэтому вместо того, чтобы накапливать счётчики вхождений, мы рассчитаем их математическое ожидание по всем сгенерированным выборкам без повторений, имеющим такое же упорядоченное множество элементов $w_i = \mathbb{E}(c_i).$

Пусть $q_i = \frac{show(v_i)}{\sum_{j=1}^{|V|} show(v_j)}$,пусть $w_i^k$ —  количество вхождений элемента с индексом $i$ к моменту, когда в выборке набралось $k$ уникальных элементов. Пусть уникальные элементы попадают в выборку в порядке $v_1, v_2, v_3, \ldots, v_n$. 

Когда мы возьмём первый элемент: 

$$
w_1^1 =1, \quad w_2^1 = 0, \quad w_3^1 = 0, \quad \ldots, \quad w_n^1 = 0.
$$

Прежде, чем мы достанем второй уникальный элемент, мы, в среднем, достанем первый элемент ещё  $\frac{1}{1 - q_1} - 1 = \frac{q_1}{1 - q_1}$ раз. 

Эту величину мы посчитали с помощью геометрического распределения. В качестве успешного события мы рассматриваем второй уникальный элемент. Вероятность успеха равна $1- q_1.$ Геометрическая случайная величина представляет из себя номер первого успешного события. Если мы хотим получить число не успешных событий, надо вычесть единицу.

Получаем

$$
w_1^2 = 1 + \frac{q_1}{1 - q_1}, \quad w_2^2 = 1, \quad w_3^2 = 0, \quad \ldots, \quad w_n^2 = 0.
$$

Прежде, чем мы достанем третий уникальный элемент, мы, в среднем, достанем первые два элемента ещё $\frac{1}{1 - (q_1 + q_2)} - 1 = \frac{q_1 + q_2}{1 - (q_1 + q_2)}$ раз. Из них доля доставания первого элемента составляет $\frac{q_1}{q_1 + q_2}$, а доля второго $\frac{q_2}{q_1 + q_2}$. 

Получаем 

$$
w_1^3 = 1 + \frac{q_1}{1 - q_1} + \frac{q_1}{1 - (q_1 + q_2)}, \quad w_2^3 = 1 + \frac{q_2}{1 - (q_1 + q_2)}, \quad  w_3^3 = 1, \quad \ldots, \quad w_n^3 = 0.
$$

Продолжая эту логику до шага $n,$ получаем формулу 

$$
c_i = w_i^n = 1 + q_i \cdot \left( \sum_{j=i}^{n-1} \frac{1}{1 - \sum_{k=1}^j q_k} \right).
$$

Подставим эти веса вместо счетчиков $c_i$. Это даст нам несмещённую оценку доли «плохих» показов на основе разметки ровно $n$ объектов

$$
\hat p_{\text{bad}} = \frac{\sum_{i=1}^n \left[ 1 + q_i \cdot \left( \sum_{j=i}^{n-1} \frac{1}{1 - \sum_{k=1}^j q_k} \right) \right] \cdot isSpam(v_i)}{\sum_{i=1}^n \left[ 1 + q_i \cdot \left( \sum_{j=i}^{n-1} \frac{1}{1 - \sum_{k=1}^j q_k} \right)  \right]},  \quad q_i = \frac{show(v_i)}{\sum_{j=1}^{|V|} show(v_j)}.
$$



__е) [1 балл]__ Пришло время закодит, полученные выше веса. Сделайте $1000$ выборок без повторений размера $1000$. Постройте для каждой оценку доли плохих показов. Убедитесь, что оценка, предложенная выше, окажется несмещённой.

**Hint:** для расчёта весов удобно воспользоваться несколько раз функцией `np.cumsum`.

In [119]:
# внимательно изучите этот код:

a = np.array([2, 3, 4, 5])
np.cumsum(a[:-1][::-1])[::-1]

array([9, 7, 4])

In [120]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz

# Параметры симуляций и выборки
n_simulations = 1000
sample_size = 1000
n_obs = df.shape[0]

# Предварительно вычисляем вероятности q для каждого видео
shows = df['shows'].values
total_shows = shows.sum()
q_all = shows / total_shows
spam_labels_all = df['is_spam'].values
p_hat_unbiased = np.empty(n_simulations)

for sim in range(n_simulations):
    # Отбираем выборку без повторений с вероятностями q_all
    sampled_indices = np.random.choice(n_obs, size=sample_size, replace=False, p=q_all)
    u_sample = np.random.uniform(size=sample_size)
    score_sample = np.log(u_sample) / shows[sampled_indices]
    
    # Сортируем выборку по убыванию score
    order = np.argsort(score_sample)[::-1]
    sorted_indices = sampled_indices[order]

    q_sampled = q_all[sorted_indices]
    spam_sampled = spam_labels_all[sorted_indices]
    
    # Вычисляем кумулятивную сумму q для отсортированной выборки
    S = np.cumsum(q_sampled)
    if sample_size > 1:
        r = 1 / (1 - S[1:])                  
        reverse_cumsum = np.cumsum(r[::-1])[::-1]
        weights = np.empty(sample_size)
        weights[:-1] = 1 + q_sampled[:-1] * reverse_cumsum  
        weights[-1] = 1.
    else:
        weights = np.array([1.])
    
    # Несмещённая оценка доли плохих показов для текущей симуляции
    p_hat_unbiased[sim] = np.sum(weights * spam_sampled) / np.sum(weights)

true_p = (df['shows'] * df['is_spam']).sum() / df['shows'].sum()
mean_unbiased = p_hat_unbiased.mean()
bias = mean_unbiased - true_p

print("Истинная доля плохих показов:", true_p)
print("Средняя несмещённая оценка:", mean_unbiased)
print("Среднее смещение:", bias)

# Строим гистограмму распределения оценок
fig = go.Figure()
fig.add_trace(go.Histogram(x=p_hat_unbiased, nbinsx=30, marker_color='skyblue'))
fig.add_vline(x=true_p, line=dict(color='red', dash='dash'),
              annotation_text=f'Истинная доля = {true_p:.3f}', annotation_position="top left")
fig.add_vline(x=mean_unbiased, line=dict(color='green', dash='dash'),
              annotation_text=f'Средняя оценка = {mean_unbiased:.3f}', annotation_position="top right")
fig.update_layout(
    title='Гистограмма несмещённых оценок доли плохих показов (1000 выборок без повторений)',
    xaxis_title='Оценка доли спама',
    yaxis_title='Частота'
)
fig.show()


Истинная доля плохих показов: 0.030451003751416496
Средняя несмещённая оценка: 0.03037719910284918
Среднее смещение: -7.380464856731467e-05


#### 6. Доверительный интервал 

Про долю плохих заказов надо уметь делать выводы. Для этого надо построить доверительный интервал. Для него нужна дисперсия. Все карточки размечаются модераторами незвисимо друг от друга. Получается, что дисперсию можно найти как 

$$
\text{Var}(\hat p_{\text{bad}}) = \frac{\sum_{i=1}^n c^2_i \cdot \text{Var}(\text{isSpam}(v_i))}{\left(\sum_{i=1}^n c_i\right)^2} = \frac{\sum_{i=1}^n c^2_i}{\left(\sum_{i=1}^n c_i\right)^2} \cdot p_{\text{bad}} \cdot (1 - p_{\text{bad}})
$$

Видно, что если сэмпл занимает небольшую долю выборки и показы распределены между элементами достаточно равномерно, то веса $c_i$ будут несильно отличаться от $1$. Тогда асимптотически, по мере роста размера сэмпла, стандартное отклонение метрики будет падать как $\frac{1}{\sqrt{n}}.$ 

Однако, если существенная доля показов в рекомендательной ленте представлена небольшой группой видосов, тогда веса будут существенно различаться и, начиная с определённого момента, увеличение размера выборки не будет давать заметного снижения разброса в оценке доли плохих показов.

__ё) [1 балл]__ Сделайте $1000$ выборок без повторений размера $1000$. Постройте по каждой $95\%$ доверительный интервал для доли плохих показов. 

Убедитесь, что он действительно покрывает долю плохих показов с вероятностью $0.95$.

In [121]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz

n_simulations = 1000
sample_size = 1000
n_obs = df.shape[0]

# Предварительные данные
shows = df['shows'].values
total_shows = shows.sum()
q_all = shows / total_shows
spam_labels_all = df['is_spam'].values
true_p = (df['shows'] * df['is_spam']).sum() / total_shows

# Массивы для хранения результатов
p_hat_unbiased = np.empty(n_simulations)
CI_lower = np.empty(n_simulations)
CI_upper = np.empty(n_simulations)

for sim in range(n_simulations):
    sampled_indices = np.random.choice(n_obs, size=sample_size, replace=False, p=q_all)
    
    u_sample = np.random.uniform(size=sample_size)
    score_sample = np.log(u_sample) / shows[sampled_indices]
    
    # Сортируем выборку по убыванию score
    order = np.argsort(score_sample)[::-1]
    sorted_indices = sampled_indices[order]
    
    q_sampled = q_all[sorted_indices]
    spam_sampled = spam_labels_all[sorted_indices]
    
    # Вычисляем кумулятивную сумму q
    S = np.cumsum(q_sampled)
    
    # Вычисляем веса
    if sample_size > 1:
        r = 1 / (1 - S[1:])                  
        reverse_cumsum = np.cumsum(r[::-1])[::-1] 
        weights = np.empty(sample_size)
        weights[:-1] = 1 + q_sampled[:-1] * reverse_cumsum
        weights[-1] = 1.
    else:
        weights = np.array([1.])
    
    # Несмещённая оценка доли спама
    hat_p = np.sum(weights * spam_sampled) / np.sum(weights)
    p_hat_unbiased[sim] = hat_p
    
    var_hat = (np.sum(weights**2) / (np.sum(weights)**2)) * hat_p * (1 - hat_p)
    se_hat = np.sqrt(var_hat)
    
    # 95% доверительный интервал (классический, симметричный)
    CI_lower[sim] = np.clip(hat_p - 1.96 * se_hat, 0, 1)
    CI_upper[sim] = np.clip(hat_p + 1.96 * se_hat, 0, 1)
coverage = np.mean((CI_lower <= true_p) & (CI_upper >= true_p))
print("Истинная доля плохих показов:", true_p)
print("Покрытие классическим ДИ:", coverage)

mean_unbiased = p_hat_unbiased.mean()

# Гистограмма ширины интервалов или распределения оценок
fig = go.Figure()
fig.add_trace(go.Histogram(x=p_hat_unbiased, nbinsx=30, marker_color='skyblue'))
fig.add_vline(x=true_p, line=dict(color='red', dash='dash'),
              annotation_text=f'Истинная доля = {true_p:.3f}', annotation_position="top left")
fig.add_vline(x=mean_unbiased, line=dict(color='green', dash='dash'),
              annotation_text=f'Средняя оценка = {mean_unbiased:.3f}', annotation_position="top right")
fig.update_layout(
    title='Гистограмма несмещённых оценок доли плохих показов (1000 симуляций)',
    xaxis_title='Оценка доли спама',
    yaxis_title='Частота'
)
fig.show()

Истинная доля плохих показов: 0.030451003751416496
Покрытие классическим ДИ: 1.0


__ж) [1 балл]__ Если вы всё сделали всё верно, выше симуляции дали очень плохой результат. Доверительный интервал развалился. Он оказался слишком широким. Более того, он пробивает слева ноль. 

Сконструируйте доверительный интервал Уилсона. Убедитесь, что он больше не пробивает отрезок $[0;1]$.

In [122]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz
n_simulations = 1000
sample_size = 1000
CI_lower_wilson = np.empty(n_simulations)
CI_upper_wilson = np.empty(n_simulations)

z = 1.96  # для 95% ДИ

for sim in range(n_simulations):
    sampled_indices = np.random.choice(n_obs, size=sample_size, replace=False, p=q_all)
    
    u_sample = np.random.uniform(size=sample_size)
    score_sample = np.log(u_sample) / shows[sampled_indices]
    
    order = np.argsort(score_sample)[::-1]
    sorted_indices = sampled_indices[order]
    
    q_sampled = q_all[sorted_indices]
    spam_sampled = spam_labels_all[sorted_indices]
    
    S = np.cumsum(q_sampled)
    if sample_size > 1:
        r = 1 / (1 - S[1:])
        reverse_cumsum = np.cumsum(r[::-1])[::-1]
        weights = np.empty(sample_size)
        weights[:-1] = 1 + q_sampled[:-1] * reverse_cumsum
        weights[-1] = 1.
    else:
        weights = np.array([1.])
    
    hat_p = np.sum(weights * spam_sampled) / np.sum(weights)
    n_eff = (np.sum(weights)**2) / np.sum(weights**2)  # эффективный размер выборки
    
    # Интервал Уилсона для биномиальной пропорции
    denominator = 1 + z**2 / n_eff
    center = (hat_p + z**2 / (2 * n_eff)) / denominator
    half_width = (z / denominator) * np.sqrt(hat_p * (1 - hat_p) / n_eff + z**2 / (4 * n_eff**2))
    
    CI_lower_wilson[sim] = np.clip(center - half_width, 0, 1)
    CI_upper_wilson[sim] = np.clip(center + half_width, 0, 1)

coverage_wilson = np.mean((CI_lower_wilson <= true_p) & (CI_upper_wilson >= true_p))
print("Покрытие ДИ Уилсона:", coverage_wilson)
print("Минимальное значение нижней границы (Уилсон):", CI_lower_wilson.min())
print("Максимальное значение верхней границы (Уилсон):", CI_upper_wilson.max())

Покрытие ДИ Уилсона: 1.0
Минимальное значение нижней границы (Уилсон): 0.0009104607041532653
Максимальное значение верхней границы (Уилсон): 0.4900018092722831


Доверительный интервал Уилсона всё еще не будет давать нужного уровня значимости из-за огромной ширины. Нам нужен какой-то способ уменьшить дисперсию. Это можно сделать с помощью машинного обучения.

#### 7. Уменьшение дисперсии с помощью машинного обучения

Когда доля маленькая, дисперсия оценки доли будет довольно высокой. Относительная ошибка в оценке быстро растёт при уменьшении доли размечаемых видосов, содержащих спам

$$
\frac{\sigma}{p_{\text{bad}}} \approx \frac{\sqrt{ p_{\text{bad}} \cdot (1 -  p_{\text{bad}})/n} }{ p_{\text{bad}} } \approx \frac{1}{\sqrt{n \cdot p_{\text{bad}}}}.
$$

Например, если реальная доля показов, содержащих спам составляет $0.5\%,$ то при разметке выборки из $1000$ видео стандартное отклонение составить $0.22\%$ и положительную метку будут обычно получать $3-7$ видео, а наша оценка доли спама будет колебаться в пределах $(0.5 \pm 0.22)\%.$ При таком уровне шума отслеживать эффект от введения различных улучшений в пайплайнах модерации становится практически невозможно. 

Однако, если бы у нас существовал способ повысить долю просэмплированных видео с положительной разметкой в сто раз, то мы бы получили оценку доли плохого в сэмпле равную $(50 \pm 1.6)\%.$ 

Принимая во внимание, что при построении сэмпла мы завысили долю плохого в сто раз, получаем, что реальная доля плохого составит  $(0.5 \pm 0.016)\%.$ То есть с помощью приоритизации мы могли бы снизить разброс примерно в $14$ раз. 

На практике мы не можем заранее угадать какие видео будут размечены как плохие. Однако у нас есть ML-модели, предсказываютщие подозрительность видео. Например, вероятность того, что видос относится к плохому классу, $\text{score}(v_i)$.

Для того, чтобы сделать размечаемую выборку более представительной по классам, но при этом сохранить репрезентативность по потоку, повысим вероятность сэмплирования видео с высоким уровнем подозрительности.

Чтобы после этого по разметке сэмпла оценить долю плохого на потоке, нам нужно обратно изменить веса видео, чтобы величины соотвествовали ожидаемым значениям без перевзвешивания. Новые веса в формуле будут равны 

$$
c_i = \sum_{i=1}^n  \frac{1 + q_i \cdot \left( \sum_{j=i}^{n-1} \frac{1}{1 - \sum_{k=1}^j q_k}\right)}{\text{score}(v_i)}, \quad q_i = \frac{\text{show}(v_i)}{\sum_{j=1}^{|V|} \text{show}(v_j)}
$$

Оценка доли будет искаться как 

$$
\hat p_{spam} = \frac{\sum_{i=1}^n c_i \cdot \text{isSpam}(v_i)}{\sum_{i=1}^n c_i}.
$$

Аналогично дисперсия такой оценки будет иметь вид 

$$
\text{Var}(\hat p_{\text{bad}}) = \frac{\sum_{i=1}^n c^2_i \cdot \text{Var}(\text{isSpam}(v_i))}{\left(\sum_{i=1}^n c_i\right)^2}.
$$

В формуле дисперсии $\text{Var}(\text{isSpam}(v_i))$ уже нельзя считать одинаковыми, так как есть зависимость между весом объекта, обусловленным $score(v_i)$ и дисперсией бернулиевской случайной величины $\text{isSpam}(v_i).$ 

Если предсказывающая модель обучена в точности на реальном распределении, наблюдаемом в потоке либо [откалибрована,](https://github.com/esokolov/ml-course-hse/blob/master/2022-fall/seminars/sem06-calibration.ipynb) то можно считать, что $\text{score}(v_i) = \mathbb{P}(\text{isSpam}(v_i) = 1).$ 

Тогда 

$$
\text{Var}(\text{isSpam}(v_i)) = \mathbb{P}(\text{isSpam}(v_i) = 1) \cdot \mathbb{P}(\text{isSpam}(v_i) = 0) = \text{score}(v_i) \cdot (1 - \text{score}(v_i)).
$$

Если откалибровать модель не представляется возможным,тогда можно оценить математическое ожидание и дисперсию элемента на основе исторических данных по разметке элементов с похожими предсказаниями модели. 

Если для какого-то объекта по каким-то причинам отсутствует $\text{score}(v_i),$ то при сэмплировании мы можем вставить ему произвольный вес и использовать верхнюю оценку на дисперсию бернуллиевской случайной величины, $\text{Var}(\text{isSpam}(v_i)) \le 0.25.$

__з) [1 балл]__  Переделайте процесс генерации данных таким образом, чтобы для них можно было обучить классификатор. С помощью любых функций из sklearn сгенерируйте датасет таким образом, чтобы спам в данных встречался в $10\%$ случаев. Обучите логистическую регрессию. Генерируйте датасет таким образом, чтобы качество модели по метрике roc-auc оказалось в районе 0.8.

* с этого момента я чуть запутался и использовал различные источники для помощи 

In [123]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split

# Генерируем датасет: 10% спама, 90% не-спама
X, y = make_classification(n_samples=10000, n_features=20, n_informative=5, n_redundant=2,
                           weights=[0.9, 0.1], class_sep=1.0, random_state=42)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

clf_lr = LogisticRegression(max_iter=1000, random_state=42)
clf_lr.fit(X_train, y_train)

# Предсказываем вероятности на тестовой выборке и оцениваем roc-auc
y_prob = clf_lr.predict_proba(X_test)[:, 1]
roc_auc = roc_auc_score(y_test, y_prob)
print("ROC AUC:", roc_auc)


ROC AUC: 0.8123000359755366


__и) [1 балл]__ Сделайте $1000$ выборок без повторений размера $1000$. Постройте по каждой $95\%$ доверительный интервал для доли плохих показов. 

Правда ли доверительный интервал стал уже? Правда ли, что он покрывает долю плохих показов с вероятностью $0.95$?

In [124]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz

n_obs_ml = 10000
np.random.seed(42)
b = 0.7
shows_ml = np.round(sts.pareto(b).rvs(n_obs_ml))

# Формируем DataFrame с данными: берем первые n_obs_ml 
df_ml = pd.DataFrame({
    'video_id': np.arange(1, n_obs_ml+1),
    'shows': shows_ml,
    'is_spam': y[:n_obs_ml] 
})
# Получаем предсказанные вероятности (score) для каждого видео с помощью обученной логистической регрессии
df_ml['score'] = clf_lr.predict_proba(X[:n_obs_ml])[:, 1]

# Вычисляем q_i для каждого видео
shows_arr = df_ml['shows'].values
total_shows_ml = shows_arr.sum()
q_all_ml = shows_arr / total_shows_ml
is_spam_all_ml = df_ml['is_spam'].values
score_all_ml = df_ml['score'].values


true_p_ml = (shows_arr * is_spam_all_ml).sum() / total_shows_ml

# Симуляция: 1000 выборок без повторений по 1000 видео
n_simulations = 1000
sample_size = 1000

p_hat_adjusted = np.empty(n_simulations)
CI_lower_adjusted = np.empty(n_simulations)
CI_upper_adjusted = np.empty(n_simulations)

for sim in range(n_simulations):
    sampled_indices = np.random.choice(n_obs_ml, size=sample_size, replace=False, p=q_all_ml)
    
    u_sample = np.random.uniform(size=sample_size)
    score_sample = np.log(u_sample) / df_ml['shows'].values[sampled_indices]
    
    # Сортируем по убыванию score_sample
    order = np.argsort(score_sample)[::-1]
    sorted_indices = sampled_indices[order]
    
    # Извлекаем для отсортированной выборки q, метки спама и предсказанный score
    q_sampled = q_all_ml[sorted_indices]
    spam_sampled = is_spam_all_ml[sorted_indices]
    score_sampled = score_all_ml[sorted_indices]
    
    S = np.cumsum(q_sampled)
    if sample_size > 1:
        r = 1 / (1 - S[1:])  
        reverse_cumsum = np.cumsum(r[::-1])[::-1]
        weights = np.empty(sample_size)
        weights[:-1] = 1 + q_sampled[:-1] * reverse_cumsum
        weights[-1] = 1.
    else:
        weights = np.array([1.])
    
    adjusted_weights = weights
    
    hat_p_spam = np.sum(adjusted_weights * spam_sampled) / np.sum(adjusted_weights)
    p_hat_adjusted[sim] = hat_p_spam
    
    var_hat = np.sum(adjusted_weights**2 * (score_sampled * (1 - score_sampled))) / (np.sum(adjusted_weights)**2)
    se_hat = np.sqrt(var_hat)
    
    CI_lower_adjusted[sim] = np.clip(hat_p - 1.96 * se_hat, 0, 1)
    CI_upper_adjusted[sim] = np.clip(hat_p + 1.96 * se_hat, 0, 1)

# Проверяем покрытие истинного значения
coverage_adjusted = np.mean((CI_lower_adjusted <= true_p_ml) & (CI_upper_adjusted >= true_p_ml))
print("Покрытие ДИ:", coverage_adjusted)

Покрытие ДИ: 1.0


__к) [1 балл]__ Проделайте то же самое для доверительного интервала Уилсона.

In [125]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz
CI_lower_wilson_adjusted = np.empty(n_simulations)
CI_upper_wilson_adjusted = np.empty(n_simulations)
z = 1.96

for sim in range(n_simulations):
    sampled_indices = np.random.choice(n_obs_ml, size=sample_size, replace=False, p=q_all_ml)
    
    u_sample = np.random.uniform(size=sample_size)
    score_sample = np.log(u_sample) / df_ml['shows'].values[sampled_indices]
    order = np.argsort(score_sample)[::-1]
    sorted_indices = sampled_indices[order]
    
    q_sampled = q_all_ml[sorted_indices]
    spam_sampled = is_spam_all_ml[sorted_indices]
    score_sampled = score_all_ml[sorted_indices]
    
    S = np.cumsum(q_sampled)
    if sample_size > 1:
        r = 1 / (1 - S[1:])
        reverse_cumsum = np.cumsum(r[::-1])[::-1]
        weights = np.empty(sample_size)
        weights[:-1] = 1 + q_sampled[:-1] * reverse_cumsum
        weights[-1] = 1.
    else:
        weights = np.array([1.])
    
    adjusted_weights = weights 
    hat_p_spam = np.sum(adjusted_weights * spam_sampled) / np.sum(adjusted_weights)
    n_eff = (np.sum(adjusted_weights)**2 / np.sum(adjusted_weights**2))
    
    denominator = 1 + z**2 / n_eff
    center = (hat_p_spam + z**2 / (2 * n_eff)) / denominator
    half_width = (z / denominator) * np.sqrt(hat_p_spam * (1 - hat_p_spam) / n_eff + z**2 / (4 * n_eff**2))
    
    CI_lower_wilson_adjusted[sim] =  np.clip(center - half_width, 0, 1)
    CI_upper_wilson_adjusted[sim] = np.clip(center + half_width, 0, 1)

coverage_wilson_adjusted = np.mean((CI_lower_wilson_adjusted <= true_p_ml) & (CI_upper_wilson_adjusted >= true_p_ml))
print("Покрытие ДИ:", coverage_wilson_adjusted)

Покрытие ДИ: 1.0


__л) [1 балл]__ Вместо логистической регрессии обучите SVM либо бустинг. Попробуйте построить доверительный интервал одним из способов, упомянутых выше. Что у вас получилось? Правда ли он покрывает долю плохих показов с вероятностью $0.95$? 

In [126]:
### ╰( ͡° ͜ʖ ͡° )つ▬▬ι═══════  bzzzzzzzzzz
from sklearn.ensemble import GradientBoostingClassifier

# Обучаем модель градиентного бустинга
clf_gb = GradientBoostingClassifier(random_state=42)
clf_gb.fit(X_train, y_train)

# Для популяции рассчитываем предсказанные вероятности по модели бустинга
df_ml['score_gb'] = clf_gb.predict_proba(X[:n_obs_ml])[:, 1]
score_all_ml_gb = df_ml['score_gb'].values

# Симуляция с использованием score_gb
p_hat_adjusted_gb = np.empty(n_simulations)
CI_lower_adjusted_gb = np.empty(n_simulations)
CI_upper_adjusted_gb = np.empty(n_simulations)

for sim in range(n_simulations):
    sampled_indices = np.random.choice(n_obs_ml, size=sample_size, replace=False, p=q_all_ml)
    
    u_sample = np.random.uniform(size=sample_size)
    score_sample = np.log(u_sample) / df_ml['shows'].values[sampled_indices]
    order = np.argsort(score_sample)[::-1]
    sorted_indices = sampled_indices[order]
    
    q_sampled = q_all_ml[sorted_indices]
    spam_sampled = is_spam_all_ml[sorted_indices]
    score_sampled_gb = score_all_ml_gb[sorted_indices]
    
    S = np.cumsum(q_sampled)
    if sample_size > 1:
        r = 1 / (1 - S[1:])
        reverse_cumsum = np.cumsum(r[::-1])[::-1]
        weights = np.empty(sample_size)
        weights[:-1] = 1 + q_sampled[:-1] * reverse_cumsum
        weights[-1] = 1.
    else:
        weights = np.array([1.])
    
    adjusted_weights = weights
    hat_p_spam =  np.sum(adjusted_weights * spam_sampled) / np.sum(adjusted_weights)
    p_hat_adjusted_gb[sim] = hat_p_spam
    var_hat = np.sum(adjusted_weights**2 * (spam_sampled - hat_p_spam)**2) / (np.sum(adjusted_weights)**2)
    se_hat = np.sqrt(var_hat)
    CI_lower_adjusted_gb[sim] = np.clip(hat_p_spam - 1.96 * se_hat, 0, 1)
    CI_upper_adjusted_gb[sim] = np.clip(hat_p_spam + 1.96 * se_hat, 0, 1)

coverage_adjusted_gb = np.mean((CI_lower_adjusted_gb <= true_p_ml) & (CI_upper_adjusted_gb >= true_p_ml))
print("Покрытие ДИ:", coverage_adjusted_gb)


Покрытие ДИ: 1.0


__м) [1 балл]__ Попробуйте в настройках эксперимента поменять долю спама в выборке с $10\%$ до $80\%$. Что произойдёт с дисперсией и смещением?



In [128]:
# Генерируем датасет с 80% спама
X80, y80 = make_classification(n_samples=10000, n_features=20, n_informative=5, n_redundant=2,
                               weights=[0.2, 0.8], class_sep=1.0, random_state=42)
from sklearn.model_selection import train_test_split
X_train80, X_test80, y_train80, y_test80 = train_test_split(X80, y80, test_size=0.3, random_state=42)

clf_lr80 = LogisticRegression(max_iter=1000, random_state=42)
clf_lr80.fit(X_train80, y_train80)
y_prob80 = clf_lr80.predict_proba(X_test80)[:, 1]
roc_auc80 = roc_auc_score(y_test80, y_prob80)
print("ROC AUC (80% spam):", roc_auc80)

# Формируем популяцию с 80% спама (используем те же shows)
df_ml80 = pd.DataFrame({
    'video_id': np.arange(1, n_obs_ml+1),
    'shows': shows_ml,
    'is_spam': y80[:n_obs_ml]
})
df_ml80['score'] = clf_lr80.predict_proba(X80[:n_obs_ml])[:, 1]

shows_arr80 = df_ml80['shows'].values
total_shows_ml80 = shows_arr80.sum()
q_all_ml80 = shows_arr80 / total_shows_ml80
is_spam_all_ml80 = df_ml80['is_spam'].values
score_all_ml80 = df_ml80['score'].values

true_p_ml80 = (shows_arr80 * is_spam_all_ml80).sum() / total_shows_ml80
print("True данные (80% spam):", true_p_ml80)

# Симуляция (классический CI с корректировкой)
p_hat_adjusted_80 = np.empty(n_simulations)
CI_lower_adjusted_80 = np.empty(n_simulations)
CI_upper_adjusted_80 = np.empty(n_simulations)

for sim in range(n_simulations):
    sampled_indices = np.random.choice(n_obs_ml, size=sample_size, replace=False, p=q_all_ml80)
    
    u_sample = np.random.uniform(size=sample_size)
    score_sample = np.log(u_sample) / df_ml80['shows'].values[sampled_indices]
    order = np.argsort(score_sample)[::-1]
    sorted_indices = sampled_indices[order]
    
    q_sampled = q_all_ml80[sorted_indices]
    spam_sampled = is_spam_all_ml80[sorted_indices]
    score_sampled = score_all_ml80[sorted_indices]
    
    S = np.cumsum(q_sampled)
    if sample_size > 1:
        r = 1 / (1 - S[1:])
        reverse_cumsum = np.cumsum(r[::-1])[::-1]
        weights = np.empty(sample_size)
        weights[:-1] = 1 + q_sampled[:-1] * reverse_cumsum
        weights[-1] = 1.
    else:
        weights = np.array([1.])
    
    adjusted_weights = weights 
    hat_p_spam = np.sum(adjusted_weights * spam_sampled) / np.sum(adjusted_weights)
    p_hat_adjusted_80[sim] = hat_p_spam
    var_hat = np.sum(adjusted_weights**2 * (score_sampled*(1 - score_sampled))) / (np.sum(adjusted_weights)**2)
    se_hat = np.sqrt(var_hat)
    CI_lower_adjusted_80[sim] =np.clip(hat_p_spam - 1.96 * se_hat, 0, 1)
    CI_upper_adjusted_80[sim] = np.clip(hat_p_spam + 1.96 * se_hat, 0, 1)

coverage_adjusted_80 = np.mean((CI_lower_adjusted_80 <= true_p_ml80) & (CI_upper_adjusted_80 >= true_p_ml80))

print("Покрытие ДИ (80% spam):", coverage_adjusted_80)
print("Средняя оценка доли спама (80% spam):", p_hat_adjusted_80.mean())
print("Стандартное отклонение оценки (80% spam):", p_hat_adjusted_80.std())


ROC AUC (80% spam): 0.8547937499999999
True данные (80% spam): 0.63639443861775
Покрытие ДИ (80% spam): 1.0
Средняя оценка доли спама (80% spam): 0.636456975379526
Стандартное отклонение оценки (80% spam): 0.0011841453252344528


__Ваш ответ:__ 
1. **ROC-AUC модели**: При увеличении доли спама до 80% качество модели (ROC-AUC) остаётся высоким (`0.855`), что указывает на хорошую разделительную способность даже при сильном дисбалансе классов.  
2. **Смещение оценки**:  
   - Средняя оценка доли спама (`0.6364`) практически совпадает с истинным значением (`0.6364`).  
   - **Смещение пренебрежимо мало**, что подтверждает адекватность метода.  
3. **Дисперсия оценки**:  
   - Стандартное отклонение оценки (`0.0012`) крайне мало, что связано с высокой долей спама в данных.  
   - Чем выше доля спама, тем меньше неопределённость в оценке, так как модель увереннее предсказывает метки.  
4. **Доверительный интервал**:  
   - Покрытие классического ДИ достигает `1.0`, что объясняется узким разбросом оценок и их близостью к истинному значению.  
   - Интервалы становятся **уже** и точнее отражают реальную долю спама.  

**Итог**:  
- При высокой доле спама (`80%`) **дисперсия оценки уменьшается**, а **смещение остаётся минимальным**.  
- Доверительные интервалы становятся более точными, но их покрытие может превышать 95% из-за крайне низкой дисперсии.  